File: 17-Binary-Tree/python/11_print_root_to_leaf_path_in_bt.py
class Solution:
    def allRootToLeaf(self, root):
        result = []

        def dfs(node, path):
            if not node:
                return

            path.append(node.data)

            if not node.left and not node.right:
                result.append(list(path))
            else:
                dfs(node.left, path)
                dfs(node.right, path)

            path.pop()

        dfs(root, [])
        return result

File: 17-Binary-Tree/cpp/11-print-root-to-leaf-path-in-bt.cpp
class Solution {
public:
    vector<vector<int>> allRootToLeaf(TreeNode* root) {
        vector<vector<int>> result;
        vector<int> path;
        dfs(root, path, result);
        return result;
    }

private:
    void dfs(TreeNode* node, vector<int>& path, vector<vector<int>>& result) {
        if (node == nullptr) return;

        path.push_back(node->data);

        if (node->left == nullptr && node->right == nullptr) {
            result.push_back(path);
        } else {
            dfs(node->left, path, result);
            dfs(node->right, path, result);
        }

        path.pop_back();
    }
};
